### ISSUE TITLE
Implement OptMeowt in Firefox

### ISSUE THREAD

**Issue by SebastianZimmeck** [privacy-tech-lab/optmeowt-browser-extension#7]
I am opening a new issue based on what @davebaraka [said earlier](https://github.com/privacy-tech-lab/privacy-rights-platform-browser-extension/issues/4):

> @kalicki1 I've only tested this chrome. Can you look at the extent this is working in Firefox. Based on issue #5, it seems we will need some way to route the correct api's for each platform. You may also want to look into polyfills like this [web extension-polyfill](https://github.com/mozilla/webextension-polyfill).

(cc'ing @rgoldstein01, @pakaelbling)

**Comment by kalicki1** [comment]
As mentioned in issue #92 by @dharb, 

>Adding a script tag to the DOM like that can be blocked by a site's CSP in Firefox (for instance, on twitter.com). Check out the section in the gist I shared about Xray vision for details.

The xray discussion was here for reference: [Setting the DOM Signal](https://gist.github.com/dharb/15bb3888c258d15cf96efec9b72dd605#setting-the-dom-signal).

It might be worthwhile to come back to this issue and fix up OptMeowt for Firefox and other browsers sometime soon given that we have progressed much farther than where we were when we first opened this issue. This will be for reference for when we get to this.

**Comment by SebastianZimmeck** [comment]
Good point, @kalicki1. Let's discuss when we meet next week.

**Comment by SebastianZimmeck** [comment]
@kalicki1 and @asalia-1 will look into how difficult it is to port OptMeowt to Firefox and also how the code can be organized (e.g., in the same repo, in different repos, with Git submodules). There is certainly a common codebase. What is the best way to handle it and keep the different versions in sync.

**Comment by asalia-1** [comment]
### To make our extension compatible across firefox, these are the key things we need to focus on in our codebase:

**Manifest.json keys** :-  There are some manifest.json keys that are not compatible with both Firefox and Chrome but luckily, most of them are compatible across both platforms. Currently, the keys in the manifest.json file is compatible accross both platforms however, there are some keys that are mandatory in Firefox that we do not have. I checked the list of manifest keys that are compatible with Chrome and Firefox and the effort needed to make our manifest.json file compatible with Chrome and Firefox is minimal. We simply have to add a few Firefox mandatory keys to achieve this.
 **API namespace** :- Generally, APIs in Chrome have the namespace chrome.* and Firefox APIs have the browser.* namespace. For a good amount of APIs this is the only difference. The good thing is that most APIs with the chrome.* namespace are also compatible with Firefox and run just fine. This explains why OptMeOwt for the seems to work somewhat well in Firefox.
 **API function coverage** :- Some Chrome APIs may simply not be available for Firefox. In such cases we would need an alternative implementation that works on Firefox
 **API asynchronous event handling** :- Another thing we would have to look at how we are handling asynchronous events. Firefox uses promises while Chrome uses Callbacks.

I found this comprehensive article that pretty much sums up the work we need to do to achieve cross platform compatibility across Chrome and Firefox: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Build_a_cross_browser_extension
In our case, I think most of the effort would be to handle APIs that simply have no equivalent in Firefox and handling asynchronous events. 
 
### Some suggestions for how we could tackle this: 
 
1.  We create an abstract Browser class that will conduct runtime checks to ensure that we expose certain APIs if we are operating in Chrome or a fallback implementation option if we are running in Firefox. In a similar vein, we can handle asynchronous events differently depending on the runtime environment using different implementations of the abstract class.
I am compiling a list of chrome APIs we use in OptMeowt to know how much this effort this will require. For some APIs, it will be as simple as changing the namespace from chrome.* to browser.*, for others it may require more implementation effort which could mean some significant alterations to our current logic flow. 
 
2.  Another method to consider is to implement the extension as targeted towards Firefox and then use the 
 [WebExtension browser API Polyfill](https://github.com/mozilla/webextension-polyfill/) library to make the extension compatible with Chrome. This was suggested in the article linked above. This is viable because the Pollyfill library takes of the issues associated with asynchronous javaScript and namespace as well. Hence if it works on Firefox, there is guarantee that it will work in Chrome using the Polyfill library. This will require a lot more effort now but may make implementation a lot more straightforward in the future

**Comment by SebastianZimmeck** [comment]
> I am compiling a list of chrome APIs we use in OptMeowt to know how much this effort this will require. 

That is good! I would say, from a practical perspective, another strategy is to run OptMeowt in its current form in Firefox and check in the Developer console where errors are occurring. Then, incrementally address those (checking again where errors are occurring once you you fixed an error). 

> Another method to consider is to implement the extension as targeted towards Firefox and then use the
WebExtension browser API Polyfill library to make the extension compatible with Chrome.

I am a bit hesitant to base the extension on Firefox since at this point Chrome/Chromium is really the majority platform. Also, as far as I know, we are using standard APIs for the most part and nothing exotic. So, I would hope that changes are relatively minor and maintainable once we have them implemented in a Firefox version (at least, I hope that is the case).

**Comment by SebastianZimmeck** [comment]
I confirmed with DuckDuckGo that their [Privacy Essentials Extension](https://github.com/duckduckgo/duckduckgo-privacy-extension) is based on the same codebase for Chrome and Firefox. They are doing some checks in the code whether it runs on Firefox and, if necessary, run some other APIs that may not be compatible with Firefox. The only other thing is separate build tasks that slightly rearrange where things go (maybe not necessary for us). As OptMeowt is probably less complex than Privacy Essentials, the DuckDuckGo approach seems suitable to me for us as well.

**Comment by SebastianZimmeck** [comment]
@dharb was so kind to offer some pointers in this thread on how DuckDuckGo went about the creation of a codebase that works for both Chrome and Firefox. @kalicki1 already started the work and will give @asalia-1 access to the code and potentially a diff.

**Comment by dharb** [comment]
Apologies for the delay here, busy week.

As you guys already mentioned, Firefox and Chrome web extensions are quite similar but there are some slight differences, especially around the build process. I'm going to link you to some files that we use for our extension and give a brief overview of what they're doing, then I think you would probably learn the fastest by cloning our extension repo, running the build tasks, and looking at the output files.

First things first, you'll want to use a different manifest for Firefox and Chrome, as there are some minor differences there. We keep these in their [own directories](https://github.com/duckduckgo/duckduckgo-privacy-extension/tree/develop/browsers). Ignore the safari directory, that's outdated and no longer used. If you diff the manifest files you should see what needs to be changed for Firefox.

Next thing to check out is how we detect which browser is being used and use that to conditionally do things. This logic can be found [here](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/background/utils.es6.js#L5) and [here](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/shared-utils/parse-user-agent-string.es6.js). To see how we use this to inject the GPC signal, take a look [here](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/background/GPC.es6.js#L26).

Now for the most complicated part: the build process. We use node/npm/grunt to build our extension. It's important to note here that we have separate build processes for dev and release, as they require different outputs. Here's how this works:

In [package.json](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/package.json) is a list of scripts that we use to build the extension. You'll notice that a lot of these commands start with `make`--any time you see that, refer to the [Makefile](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/Makefile) to see what that actually does. Using make like this is great for doing things that you would typically do in bash, eg creating directories and moving files around.

The part you should focus on here are the specific build commands: `dev-firefox` `dev-chrome` `release-firefox` `release-chrome`. These essentially process things and output to a browser-specific directory: `/build/firefox` or `/build/chrome`.

I realize this is pretty dense, so if I were you I would run through the whole build process with our extension and check out the output files:
- `git clone https://github.com/duckduckgo/duckduckgo-privacy-extension.git`
- `cd duckduckgo-privacy-extension && npm i`
- `npm run dev-chrome` (these commands leave the build process monitoring files for changes)
- `npm run dev-firefox`
- `npm run release-chrome`

I hope this is a helpful start. Feel free to reach out if you need help, or something doesn't make sense.
- `npm run release-firefox`

**Comment by asalia-1** [comment]
Thanks @dharb ! This is very helpful

**Comment by SebastianZimmeck** [comment]
Indeed, thank you, @dharb! At the moment, we are not using any build process. However, as I understand that this process is the basis for creating the different versions of the extension, this seems to be the route for us as well.

@asalia-1, can you make some headway on this by Tuesday?

@kalicki1, does the general approach via the build process fit into what you have done so far above?

**Comment by asalia-1** [comment]
Making changes to the manifest file and conditionally calling functions depending on the browser user agent is straight forward and I can get that done over the weekend. Currently we do not have npm/grunt integrated with our extension so that will be a bit of unknown territory pertaining the build process but I will reach out to David if I have any questions about that.

**Comment by SebastianZimmeck** [comment]
@asalia-1, great! Feel free to open additional issues (if you think it is warranted; otherwise, just working in this one is fine as well).

**Comment by SebastianZimmeck** [comment]
Please feel free to close this issue, @asalia-1.

**Comment by SebastianZimmeck** [comment]
@kalicki1 will look into the [semantic versioning](https://semver.org/) for the new Firefox-included version.

**Comment by SebastianZimmeck** [comment]
We have a privacy-tech-lab account on https://addons.mozilla.org/en-US/firefox/ now to publish the packaged extension.

**Comment by kalicki1** [comment]
I updated the tag for our first Firefox-compatible release as `v1.1` based on our discussions and the semantic versioning @SebastianZimmeck mentioned. Given that we have resolved this issue, and that we will most likely discuss releasing to the Firefox addons site soon, I will go ahead and close this issue.
